### ISSUE TITLE
Issue creating persistent storage for /var/lib/awx/projects in awx-task container

### ISSUE THREAD

**Issue by jamesfreeman959** [ansible/tower-operator#259]
Hi there

I have deployed AWX via awx-operator 0.8.0 on microk8s 1.20.6 on Ubuntu 20.04 LTS. All seems to be working well except that I cannot figure out how to get the /var/lib/awx/projects directory to live in a persistent storage volume. Right now this is a simple, single node setup, so I have created a persistent volume claim as follows:

```
apiVersion: v1
kind: PersistentVolumeClaim
metadata:
  name: awx-pvc
spec:
  accessModes:
    - ReadWriteMany
  storageClassName: microk8s-hostpath
  resources:
    requests:
      storage: 1Gi
```

I've been playing around with volumes, but my current deployment description for AWX looks like this:

```
apiVersion: awx.ansible.com/v1beta1
kind: AWX
metadata:
  name: awx
spec:
  tower_ingress_type: Ingress
  tower_ingress_tls_secret: awx-secret-tls
  tower_hostname: awx.example.org
  tower_projects_existing_claim: awx-pvc
  tower_extra_volumes: | 
    - name: "data-vol"
      hostPath: 
        path: "/var/lib/awx"
        type: Directory 
  tower_task_extra_volume_mounts: | 
    - name: "data-vol" 
      mountPath: "/data"
```

All pods come up fine, and I can log into AWX just fine - however whenever I go to create a new Project with Source Control Type Manual, it says that there are not available playbook directories in /var/lib/awx/projects, even if I manually copy a playbook directory into the persistent volume I created earlier.

Happy to supply any and all debugging information for this - from the documentation I wasn't really clear how to achieve this, but I had high hopes for "tower_projects_existing_claim". 

Any help appreciated!

Many thanks

James

**Comment by tchellomello** [comment]
Hello @jamesfreeman959, I think what is missing you to pass the option below:

```yaml
apiVersion: awx.ansible.com/v1beta1
kind: AWX
metadata:
  name: awx
spec:
  tower_ingress_type: Ingress
  tower_ingress_tls_secret: awx-secret-tls
  tower_hostname: awx.example.org
  tower_projects_existing_claim: awx-pvc
  tower_projects_persistence: true  <--- this
```

For example:
```yaml
apiVersion: awx.ansible.com/v1beta1
kind: AWX
metadata:
  labels:
    app.kubernetes.io/component: awx
    app.kubernetes.io/managed-by: awx-operator
    app.kubernetes.io/name: awx-toca
    app.kubernetes.io/part-of: awx-toca
  name: awx-toca
  namespace: default
spec:
  ldap_cacert_secret: awx-toca-ldap-ca-cert
  tower_admin_email: tchello.mello@gmail.com
  tower_admin_password_secret: awx-toca-admin-password
  tower_admin_user: admin
  tower_create_preload_data: true
  tower_ee_extra_volume_mounts: |
    - name: ansible-cfg
      mountPath: /etc/ansible/ansible.cfg
      subPath: ansible.cfg
  tower_extra_volumes: |
    - name: ansible-cfg
      configMap:
        defaultMode: 420
        items:
          - key: ansible.cfg
            path: ansible.cfg
        name: awx-toca-extra-config
    - name: custom-py
      configMap:
        defaultMode: 420
        items:
          - key: custom.py
            path: custom.py
        name: awx-toca-extra-config
  tower_garbage_collect_secrets: true
  tower_hostname: awx.tatu.home
  tower_image_pull_policy: IfNotPresent
  tower_ingress_type: none
  tower_projects_persistence: true
  tower_projects_storage_access_mode: ReadWriteMany
  tower_projects_storage_size: 8Gi
  tower_replicas: 1
  tower_task_extra_volume_mounts: |
    - name: custom-py
      mountPath: /etc/tower/conf.d/custom.py
      subPath: custom.py
  tower_task_privileged: false
  tower_task_resource_requirements:
    limits:
      cpu: 1000m
      memory: 2Gi
    requests:
      cpu: 500m
      memory: 1Gi
  tower_web_resource_requirements:
    limits:
      cpu: 2000m
      memory: 4Gi
    requests:
      cpu: 1000m
      memory: 2Gi
```
Results in this:

```yaml

kubectl get pvc awx-toca-projects-claim                                              23:52:15
NAME                      STATUS   VOLUME                                     CAPACITY   ACCESS MODES   STORAGECLASS   AGE
awx-toca-projects-claim   Bound    pvc-28bfde43-e1e2-44ba-8863-9b4e1f3d1d26   8Gi        RWX            nfs-scratch    47h


kubectl iexec awx-toca /bin/bash                                                                                                         23:53:05
Container: ✔ awx-toca-task
bash-4.4$ df -h /var/lib/awx/projects
Filesystem                                                                                                    Size  Used Avail Use% Mounted on
192.168.111.46:/volume1/k8s_scratch/default-awx-toca-projects-claim-pvc-28bfde43-e1e2-44ba-8863-9b4e1f3d1d26  3.5T  1.9T  1.7T  52% /var/lib/awx/projects

```

 You don't need to use the `extra_volume_mount` as the `/var/lib/awx/projects`  should be cover that using the options above. 

 Check the https://github.com/ansible/awx-operator#persisting-projects-directory and https://github.com/ansible/awx-operator/pull/177 for further information.

**Comment by jamesfreeman959** [comment]
Hi @tchellomello 

Thanks so much for your awesome help - so progress has been made. I added the missing line to the YAML file defining the `my-awx` service, and whilst I'm not using NFS, I'm just using a local disk store (single node, super simple example), everything looks better:

```
$ microk8s kubectl get pods
NAME                          READY   STATUS    RESTARTS   AGE
awx-operator-f768499d-njnmq   1/1     Running   1          19h
awx-postgres-0                1/1     Running   0          14m
awx-86d66d84b-f9rdg           4/4     Running   0          13m

$ microk8s kubectl exec --stdin --tty awx-86d66d84b-f9rdg --container awx-task -- /bin/bash

bash-4.4$ cd /var/lib/awx/projects/
bash-4.4$ ls -l
total 4
drwxrwxr-x 2 awx  1000 4096 Apr 29 16:24 test
bash-4.4$ cd test
bash-4.4$ ls -l
total 4
-rw-rw-r-- 1 awx  1000   31 Apr 29 16:24 site.yml
```

So I can see my test file in there, which I know I created within the PVC on my microk8s host, so the persistence is definitely working. This is awesome!

However, just to throw a spanner into the works, when I log into the AWX GUI and go to add a project of type "manual", it says:

```
WARNING:
There are no available playbook directories in /var/lib/awx/projects. Either that directory is empty, or all of the contents are already assigned to other projects. Create a new directory there and make sure the playbook files can be read by the "awx" system user, or have AWX directly retrieve your playbooks from source control using the Source Control Type option above.
```

The playbook file is there and has become owned by `awx` within the awx-task container, so I think everything is ok. Have I missed something else that would prevent the GUI from finding the test playbooks directory?

Thanks for all your help!

James

**Comment by tchellomello** [comment]
Great news @jamesfreeman959 

> 
> However, just to throw a spanner into the works, when I log into the AWX GUI and go to add a project of type "manual", it says:
> 
> ```
> WARNING:
> There are no available playbook directories in /var/lib/awx/projects. Either that directory is empty, or all of the contents are already assigned to other projects. Create a new directory there and make sure the playbook files can be read by the "awx" system user, or have AWX directly retrieve your playbooks from source control using the Source Control Type option above.
> ```
> 
> The playbook file is there and has become owned by `awx` within the awx-task container, so I think everything is ok. Have I missed something else that would prevent the GUI from finding the test playbooks directory?
>
> Thanks for all your help!
> 
> James

 For that, make sure you have the `/var/lib/awx/projects/mydir` and then the playbook inside of that. Makes sure the directory is also owned by the `awx` user.

**Comment by jamesfreeman959** [comment]
Thanks @tchellomello - unfortunately I have that set up (though if you can see an error in my setup please let me know!).

This output was taken from a shell inside the `awx-task` container:

```
bash-4.4$ cd /var/lib/awx/projects/
bash-4.4$ find
.
./deploy
./deploy/site.yml
bash-4.4$ cat deploy/site.yml 
---
- name: Test
  hosts: all

  tasks:
    - name: Print a debug message
      debug:
        msg: "Hello World"
bash-4.4$ ls -l
total 4
drwxrwxr-x 2 awx 1000 4096 Apr 30 11:45 deploy
bash-4.4$ ls -l deploy/
total 4
-rw-rw-r-- 1 awx 1000 114 Apr 30 11:45 site.yml
```

As you can see, I have a valid (albeit Hello World) playbook, the file itself and directory are owned by `awx` and are under `/var/lib/awx/projects`.

Am I missing something or is something invalid in my setup that would prevent AWX from seeing these?

**Comment by tchellomello** [comment]
Interesting.. I gave it a try and yes, I was able to reproduce it. 

```yaml
kubectl iexec awx-toca /bin/bash                                                           13:23:47
Container: ✔ awx-toca-task
bash-4.4$ namei  -lm /var/lib/awx/projects/manual/site.yml 
f: /var/lib/awx/projects/manual/site.yml
drwxr-xr-x root root /
drwxr-xr-x root root var
drwxr-xr-x root root lib
drwxrwxr-x root root awx
drwxrwxrwx root root projects
drwxr-xr-x awx  root manual
-rw-r--r-- awx  root site.yml
```

![image](https://user-images.githubusercontent.com/809840/116731187-61cf0a80-a9b7-11eb-8511-6b1a83750773.png)


So the problem here is because the `awx-web` does not mount the  `/var/lib/projects` and that is a requirement when setting it manually.  Once I created it on the `awx-web` it worked like a charm. 

![image](https://user-images.githubusercontent.com/809840/116731544-d013cd00-a9b7-11eb-8428-46df37a30ec1.png)

Once I modified the operator code with the change below, then it worked out of the box. 

Created PR https://github.com/ansible/awx-operator/pull/263
